### PR TITLE
add jason writer class

### PR DESCRIPTION
**Pull request by jafu888**
This code copies out the JASON writing logic from ConstraintSet.java.
Eventually we can delete it from ConstraintSet.java.

### CODE REVIEW

File: projects/MotionLayoutVerification/app/src/main/java/android/support/constraint/app/MotionLayoutToJason.java [hunk 1]
```diff
@@ -0,0 +1,430 @@
...
+        WriteJsonEngine(Writer writer, ConstraintSet set, ConstraintLayout layout, int flags) throws IOException {
+            this.mWriter = writer;
+            this.mLayout = layout;
+            this.mContext = layout.getContext();
+            this.mFlags = flags;
+            this.set = set;
+            set.getConstraint(2);
+        }
+
+        int []getIDs() {
```
Comment (jswong65) [inline projects/MotionLayoutVerification/app/src/main/java/android/support/constraint/app/MotionLayoutToJason.java:89]: nit: int[] getIDs()

File: projects/MotionLayoutVerification/app/src/main/java/android/support/constraint/app/MotionLayoutToJason.java [hunk 2]
```diff
@@ -0,0 +1,430 @@
...
+        private static final String SPACE = "    ";
+        private static final String SP = "  ";
+
+        WriteJsonEngine(Writer writer, ConstraintSet set, ConstraintLayout layout, int flags) throws IOException {
+            this.mWriter = writer;
+            this.mLayout = layout;
+            this.mContext = layout.getContext();
+            this.mFlags = flags;
+            this.set = set;
+            set.getConstraint(2);
```
Comment (jswong65) [inline projects/MotionLayoutVerification/app/src/main/java/android/support/constraint/app/MotionLayoutToJason.java:86]: Do we need this?

File: projects/MotionLayoutVerification/app/src/main/java/android/support/constraint/app/MotionLayoutToJason.java [hunk 3]
```diff
@@ -0,0 +1,430 @@
...
+       ConstraintSet.Constraint getConstraint(int id)  {
+            return set.getConstraint(id);
+        }
+
+        void writeLayout() throws IOException {
+            mWriter.write( "{\n");
+            for (Integer id : getIDs()) {
+                ConstraintSet.Constraint c = getConstraint(id);
+                String idName = getSimpleName(id);
+                mWriter.write(SP+idName + ":{\n");
```
Comment (jswong65) [inline projects/MotionLayoutVerification/app/src/main/java/android/support/constraint/app/MotionLayoutToJason.java:102]: nit: SP+idName add space in-between

File: projects/MotionLayoutVerification/app/src/main/java/android/support/constraint/app/MotionLayoutToJason.java [hunk 4]
```diff
@@ -0,0 +1,430 @@
...
+        int mFlags;
+        int mUnknownCount = 0;
+        final String mLEFT = "left";
+        final String mRIGHT = "right";
+        final String mBASELINE = "baseline";
+        final String mBOTTOM = "bottom";
+        final String mTOP = "top";
+        final String mSTART = "start";
+        final String mEND = "end";
+        private static final String SPACE = "    ";
```
Comment (jswong65) [inline projects/MotionLayoutVerification/app/src/main/java/android/support/constraint/app/MotionLayoutToJason.java:77]: nit: Maybe SPACE -> FOUR_SPACES, SP -> DOUBLE_SPACES

File: projects/MotionLayoutVerification/app/src/main/java/android/support/constraint/app/MotionLayoutToJason.java [hunk 5]
```diff
@@ -0,0 +1,430 @@
...
+
+        void writeLayout() throws IOException {
+            mWriter.write( "{\n");
+            for (Integer id : getIDs()) {
+                ConstraintSet.Constraint c = getConstraint(id);
+                String idName = getSimpleName(id);
+                mWriter.write(SP+idName + ":{\n");
+                ConstraintSet.Layout l = c.layout;
+                if (l.mReferenceIds != null) {
+                    String ref = "type: '_"+idName+"_' , contains: [";
```
Comment (jswong65) [inline projects/MotionLayoutVerification/app/src/main/java/android/support/constraint/app/MotionLayoutToJason.java:105]: nit: '_"+idName+"_' add space in-between

File: projects/MotionLayoutVerification/app/src/main/java/android/support/constraint/app/MotionLayoutToJason.java [hunk 6]
```diff
@@ -0,0 +1,430 @@
...
+        private void writeMotion(ConstraintSet.Motion motion) {
+        }
+
+        private void writeTransform(ConstraintSet.Transform transform) {
+            
+        }
+
+        void writeCustom( HashMap<String, ConstraintAttribute> cset ) throws IOException {
+            if (cset != null && cset.size() > 0) {
+                mWriter.write(SPACE+"custom: {\n");
```
Comment (jswong65) [inline projects/MotionLayoutVerification/app/src/main/java/android/support/constraint/app/MotionLayoutToJason.java:186]: nit: SPACE+"custom  Add space in-between

File: projects/MotionLayoutVerification/app/src/main/java/android/support/constraint/app/MotionLayoutToJason.java [hunk 7]
```diff
@@ -0,0 +1,430 @@
...
+                        ref += ((r == 0) ? "" : ", ") + getName(rid);
+                    }
+                    mWriter.write(ref + "]\n");
+                }
+                if (l.mReferenceIdString != null) {
+                    String ref = SP + "type: '???' , contains: [";
+                    String []rids = l.mReferenceIdString.split(",");
+                    for (int r = 0; r < rids.length; r++) {
+                        String rid = rids[r];
+                        ref += ((r == 0) ? "" : ", ") + "`" +rid +"`";
```
Comment (jswong65) [inline projects/MotionLayoutVerification/app/src/main/java/android/support/constraint/app/MotionLayoutToJason.java:117]: nit: +rid + -> add space before rid

File: projects/MotionLayoutVerification/app/src/main/java/android/support/constraint/app/MotionLayoutToJason.java [hunk 8]
```diff
@@ -0,0 +1,430 @@
...
+                if (l.mReferenceIds != null) {
+                    String ref = "type: '_"+idName+"_' , contains: [";
+                    for (int r = 0; r < l.mReferenceIds.length; r++) {
+                        int rid = l.mReferenceIds[r];
+                        ref += ((r == 0) ? "" : ", ") + getName(rid);
+                    }
+                    mWriter.write(ref + "]\n");
+                }
+                if (l.mReferenceIdString != null) {
+                    String ref = SP + "type: '???' , contains: [";
```
Comment (jswong65) [inline projects/MotionLayoutVerification/app/src/main/java/android/support/constraint/app/MotionLayoutToJason.java:113]: Just curious when we would have the type: '???'

File: projects/MotionLayoutVerification/app/src/main/java/android/support/constraint/app/MotionLayoutToJason.java [hunk 9]
```diff
@@ -0,0 +1,430 @@
...
+            
+        }
+
+        void writeCustom( HashMap<String, ConstraintAttribute> cset ) throws IOException {
+            if (cset != null && cset.size() > 0) {
+                mWriter.write(SPACE+"custom: {\n");
+                for (String s : cset.keySet()) {
+
+                    ConstraintAttribute attr = cset.get(s);
+                    String custom = SPACE+SP+attr.getName() +": ";
```
Comment (jswong65) [inline projects/MotionLayoutVerification/app/src/main/java/android/support/constraint/app/MotionLayoutToJason.java:190]: nit: Add space in-between

**Review by jswong65 — COMMENTED** [review]
Thanks for helping working on the JSON writer class. Added some comments (mostly are about the formatting). I think a unit test for showing the conversion from XML (input) to JSON (expected output) could be useful, but we could do it later as a follow up PR.

File: projects/MotionLayoutVerification/app/src/main/java/android/support/constraint/app/MotionLayoutToJason.java (hunk 1 above)
Comment (jafu888) [reply projects/MotionLayoutVerification/app/src/main/java/android/support/constraint/app/MotionLayoutToJason.java]: done

File: projects/MotionLayoutVerification/app/src/main/java/android/support/constraint/app/MotionLayoutToJason.java (hunk 3 above)
Comment (jafu888) [reply projects/MotionLayoutVerification/app/src/main/java/android/support/constraint/app/MotionLayoutToJason.java]: done

File: projects/MotionLayoutVerification/app/src/main/java/android/support/constraint/app/MotionLayoutToJason.java (hunk 7 above)
Comment (jafu888) [reply projects/MotionLayoutVerification/app/src/main/java/android/support/constraint/app/MotionLayoutToJason.java]: done

File: projects/MotionLayoutVerification/app/src/main/java/android/support/constraint/app/MotionLayoutToJason.java [hunk 10]
```diff
@@ -0,0 +1,430 @@
...
+                            custom += attr.getIntegerValue();
+                            break;
+                        case COLOR_TYPE:
+                            custom += colorString(attr.getColorValue());
+                            break;
+                        case FLOAT_TYPE:
+                            custom += attr.getFloatValue();
+                            break;
+                        case STRING_TYPE:
+                            custom += "'"+attr.getStringValue()+"'";
```
[the remark this answers is not in the document]
Comment (jafu888) [reply projects/MotionLayoutVerification/app/src/main/java/android/support/constraint/app/MotionLayoutToJason.java]: done

File: projects/MotionLayoutVerification/app/src/main/java/android/support/constraint/app/MotionLayoutToJason.java [hunk 11]
```diff
@@ -0,0 +1,430 @@
...
+                        case REFERENCE_TYPE:
+                        case COLOR_DRAWABLE_TYPE:
+                        case BOOLEAN_TYPE:
+                            custom = null;
+                    }
+                    if (custom != null) {
+                        mWriter.write(custom + "\n");
+                    }
+                }
+                mWriter.write( SP+"   } \n");
```
[the remark this answers is not in the document]
Comment (jafu888) [reply projects/MotionLayoutVerification/app/src/main/java/android/support/constraint/app/MotionLayoutToJason.java]: done

File: projects/MotionLayoutVerification/app/src/main/java/android/support/constraint/app/MotionLayoutToJason.java (hunk 2 above)
Comment (jafu888) [reply projects/MotionLayoutVerification/app/src/main/java/android/support/constraint/app/MotionLayoutToJason.java:122]: remove

File: projects/MotionLayoutVerification/app/src/main/java/android/support/constraint/app/MotionLayoutToJason.java [hunk 12]
```diff
@@ -0,0 +1,430 @@
...
+import java.io.Writer;
+import java.util.HashMap;
+
+public class MotionLayoutToJason {
+    MotionLayout mMotionLayout;
+    private String TAG = "ML_DEBUG";
+
+    public void setMotionLayout(MotionLayout motionLayout) {
+        mMotionLayout = motionLayout;
+        int[] mid = motionLayout.getConstraintSetIds();
```
[the remark this answers is not in the document]
Comment (jafu888) [reply projects/MotionLayoutVerification/app/src/main/java/android/support/constraint/app/MotionLayoutToJason.java:28]: change to id

**Pull request merged**